### DISCUSSION TITLE
Add support for Ideapad (and maybe Yoga)

### DISCUSSION THREAD

**Discussion by mooms06** [BartoszCichecki/LenovoLegionToolkit#1429] [Ideas]
Currently the only missing feature in LenovoLegionToolkit  for complete support of the Ideapad Pro 5 16APH8 is the CPU power switch, every other features are working.

I found an other open source utility on Github who have the missing feature: [IdeapadToolkit](https://github.com/reagcz/IdeapadToolkit).
But it need .Net 6.0, and it does not support hotkey switching, so it's not a perfect repplacement to Vantage, like LLT is.

Maybe you can join your forces to make the ultimate app, or maybe some dev can add the support to LenovoLegionToolkit with the help of the sources of IdeapadToolkit ?

**Comment by mooms06** [comment]
As I said, the only missing features is switching the CPU power modes, I can gladly help with the tests on my own device.

**Comment by BartoszCichecki** [comment]
As I said many times - I will not add support for Ideapads/Yoga/etc because I do not have the time or resources. If someone wants to add support, I open for it.

> found an other open source utility on Github who have the missing feature: [IdeapadToolkit](https://github.com/reagcz/IdeapadToolkit).

Respectfully, this project has nothing that is remotely interesting, as it doesn't implement the feature, but require a DLL from Vantage to be copied. Which is neither a good solution nor it is compliant with the EULA of Lenovo's software.

**Comment by mooms06** [comment]
Yes, I know you said that, but as it stands, LLT already almost fully supports my Ideapad, the one missing function being the TDP limit switch.
LLT could be an universal replacement for Vantage...

I'm not a dev myself, otherwise I probably would already implemented it.  Speaking of which, how hard would it be ?  Do we need to "spy" the communications between Vantage and the laptop to be able to replicate it in LLT ? I'm not a dev but I'm a power user, not afraid to run some command lines or such, but I need some guidance...

Thanks for your answer.
Regards.

**Reply by BartoszCichecki** [reply]
What is needed most likely is getting [Ghidra](https://github.com/NationalSecurityAgency/ghidra) and reverse engineering the DLL that the project mentioned above uses. [WMIExplorer ](https://github.com/vinaypamnani/wmie2?tab=readme-ov-file) might also be handy since most likely these power modes are changed through one of the methods on one of the WMI classes starting with `LENOVO_` - they are registered under `root\WMI`.

**Reply by BartoszCichecki** [reply]
Another note: the other likely option is that these power modes are changed via energy driver, which will be more tricky.

**Comment by mooms06** [comment]
I'm afraid that's above my competences.

I've launched WMIExplorer and made a filter with "Lenovo",  here are the results, do you think it can help ? I have no idea what I'm doing.

![image](https://github.com/user-attachments/assets/c0488a5f-616b-4cf4-bf30-eaedebdad32e)

**Reply by BartoszCichecki** [reply]
That is the tricky part - you need to go through most of these and see if there is a method that sounds like it could be "it". Or going the Ghidra route is also an option.

**Comment by mooms06** [comment]
[IdeapadToolkit](https://github.com/reagcz/IdeapadToolkit) has been updated to .Net 8.0. 
Coupled with LenovoLegionToolkit, it's a complete replacement to Vantage, at least on my IdeaPad.

**Comment by HackintoshwithUbuntu** [comment]
Just in case anybody is looking to try work on this in the future, here is the (slight) progress I made. On my IdeaPad Pro 5 Gen 10 AMD the IdeapadToolkit software was not working even with the dll provided.

I did try the ghidra route, but to be honest that level of reverse engineering seemed beyond my skillset - though some of the .NET dlls with items like contract information were able to decompile quite nicely with other tools. 

I tried using WMIExplorer as suggested above (you need to use administrator to see instance information) and combined with some WMI logging it looks like Vantage uses LENOVO_OTHER_METHOD::SetFeatureValue for setting the power mode, however I didn't have much luck determining what ID it uses to set the mode. Since its a uint32 I imagine the required id likely one of the hex items that can be seen in Ghidra inside powerbattery.dll. I tried brute forcing the first 2048 with a powershell script using LENOVO_OTHER_METHOD::GetFeatureValue (which only needs and ID and not a value) but didn't have any luck.
